### ISSUE TITLE
New rule: H32 - submit-button

### ISSUE THREAD

**Issue by jmucha** [dequelabs/axe-core#529]
#  H32 -  submit-button

The test checks, if data in a form can be submitted explicitly by using submit buttons (input type="submit", input type="image", or button type="submit").

**Tags:** wcag2a, wcag322, cat.forms, experimental

## Selector

Select each node that has an `form` element with `action` and `method` attributes

## forms/submit-button
- Check if the form contains a button [one of (input type="submit", input type="image", or button type="submit"]; Pass if true.
- Check if only `textarea` element(s) are present with no buttons; Forms don't have input:not(hidden) or select elements; Fail if true.
- Otherwise, cantTell

**Comment by WilcoFiers** [comment]
I think in our discussions we decided on limiting this to `form` elements with `action` and `method` attributes.

@goodwitch can you help review this one? There was recently a discussion in JC about a related issue, right?

**Comment by goodwitch** [comment]
@WilcoFiers the key words to focus on...are "when the form can be submitted implicitly".   See the discussion (from the Jedi Council) documented here https://bitbucket.org/dmusser/deque-way/pull-requests/110/consensus-jedi-decision-on-simple-search/diff#comment-44684588

**Comment by WilcoFiers** [comment]
@goodwitch I'm not clear on what these conclusions mean. Is it correct to conclude that forms without a submit button should not be considered accessibility violations? Are there ever cases where that wouldn't be true?

**Comment by goodwitch** [comment]
@WilcoFiers forms that can be submitted implicitly are not a failure of WCAG 2.0 A/AA.   Here is a description of when a form cannot be implicitly submitted (and when it can be implicitly submitted):

> If the form has no submit button, then the implicit submission mechanism **must do nothing** if the form has more than one field that blocks implicit submission, and must submit the form element from the form element itself otherwise.

**Comment by jmucha** [comment]
I wonder what if scripting is off. The page should work without client-side scripting too.

@WilcoFiers good point. We talked about limiting it.

**Comment by WilcoFiers** [comment]
So @goodwitch raises the concern that in certain cases forms with implicit submit mechanisms do meet the accessibility requirements. I think maybe we should do the following:

1. Raise a violation if the form contains only `textarea` elements, since pressing enter on those won't submit the form
2. Have the result be incomplete in all other cases

We should also exclude forms that don't contain any visible form controls.

**Comment by jmucha** [comment]
I've tweaked it a bit. 
Let's check first if the form contains not only textarea elements. Then we could proceed to looking for buttons.
In case of not fulfilling any criteria, the test will return the incomplete status.

**Comment by WilcoFiers** [comment]
@jmucha I don't think the way this check is described now works. Checks don't execute conditionally. They'll all run and their results are aggregated. I think that's what you mean, but that's not quite what the description says.

I'd suggest just putting this all into a single check. Forms pass if they have a button, forms fail if they have no button and have `textarea`(s) but don't have `input:not(hidden)` or `select` elements, otherwise, cantTell.

**Comment by jmucha** [comment]
Ok, we can put it into one check.

**Comment by dylanb** [comment]
There is no instance where this rule can definitively determine that a violation exists.

Therefore I think this rule should be inverted. It should pass forms with a valid submit mechanism and it should pass forms that meet the implicit submission restrictions. In all other cases it should return a review item.

**Comment by dylanb** [comment]
I think this would be a noisy rule and we should close this @WilcoFiers

**Comment by WilcoFiers** [comment]
I agree with Dylan. Potential for false positives on this is very high, if we make it all "needs review" it's just going to create noise.
